cps/services/worker.py

```diff
@@ -231,7 +231,8 @@ def error(self, x):
 
     @property
     def runtime(self):
-        return (self.end_time or datetime.now()) - self.start_time
+        now = datetime.now(timezone.utc)
+        return (self.end_time or now) - self.start_time
```

Comment:
Are start_time and end_time exclusively used to compute the elapsed time in seconds?

If so, you don't need to use datetime objects, which are complex classes meant to represent calendar dates and localized time.

Internally, Linux and Windows represent time in (fractions of) seconds since a certain epoch. This system clock is the source for datetime objects, but you can also use it directly whenever you want to compute the delta:

```pycon
>>> import time
>>> start_time = time.time()
>>> start_time
1720312378.525805
>>> type(start_time)
<class 'float'>
>>>
>>> end_time = time.time()
>>> end_time
1720312401.982123
>>> elapsed = end_time - start_time
>>> elapsed
23.456317901611328
>>> print(f"This example took exactly {elapsed} seconds to type, but we don't know when it started and ended in your timezone.")
This example took exactly 23.456317901611328 seconds to type, but we don't know when it started and ended in your timezone.
```

Reply:
Important caveat from the docs of [time.time()](https://docs.python.org/3/library/time.html#time.time):

> While this function normally returns non-decreasing values, it can return a lower value than a previous call if the system clock has been set back between the two calls.

This problem also affects `datetime.datetime.now(datetime.timezone.utc)` which you're using now, because the time can jump due to NTP, daylight saving or the system administrator simply setting the clock.

A more robust way to compute time differences is using `time.monotonic()`, which exists since Python 3.3:

```
>>> start_time = time.monotonic()
>>> end_time = time.monotonic()
>>> elapsed = end_time - start_time
>>> elapsed
2.559895193000557
>>> type(elapsed)
<class 'float'>
```

Reply:
1) Very useful knowledge: thanks @codewiz.

2) Maybe everyone sleep on this overnight to consider whether `datetime.now(timezone.utc)` **IS-OR-IS-NOT** good enough for these use cases ?

Reply:
> time can jump due to [1] NTP, [2] daylight saving or [3] the system administrator simply setting the clock

@deldesir above is good reminder that mandating UTC is a "step forward" — _but possibly might not be quite enough... ?_

- Mandating UTC eliminates 1 of 3 ways that time/clock can go backwards (protection against daylight savings time changes!) ✅
- BUT, the other 2 ways that time/clock can go backwards (NTP and sysadmin simply setting the clock) remain risks/considerations we need to at least think through — do we really want to live with these 2 risks — _or is `time.monotonic()` (or similar) better for some of these uses cases?_ ⚠️

Reply:
> Are start_time and end_time exclusively used to compute the elapsed time in seconds?

No. They're mainly to render the localized task status based  on the runtime per https://github.com/iiab/calibre-web/blob/master/cps/tasks_status.py

> If so, you don't need to use datetime objects, which are complex classes meant to represent calendar dates and localized time.
> 
> Internally, Linux and Windows represent time in (fractions of) seconds since a certain epoch. This system clock is the source for datetime objects, but you can also use it directly whenever you want to compute the delta:

I agree elapsed time computation can and must be simplified, however we're dealing with a codebase in which mainly naive datetime objects are used. Just lookup `datetime.nowutc` and see for yourself.

Reply:
> * Mandating "UTC as perceived by the local machine" when recording time snapshots... does eliminate 1 of 3 ways that time/clock can go backwards (i.e. protecting against impromptu daylight savings time changes!) ✅

Also be aware of a 4th way IIAB system clock can change rather automatically:

- https://github.com/iiab/iiab-admin-console/pull/204#issuecomment-493790028

(e.g. when anybody browses from a nearby phones/devices in a school!)

Summary of above PR from May 2019:

> 1. the [apache_allow_sudo](https://github.com/iiab/iiab-admin-console/blob/47d0da82ceac2742a7f91422645c600c5bf41d6c/roles/js-menu/tasks/main.yml#L131-L137) flag must be true (set with local vars) and the menu config must have [allow_server_time_update](https://github.com/search?q=repo%3Aiiab%2Fiiab-admin-console%20allow_server_time_update&type=code) true (set through the admin console)
>
> 2. if a device is mobile, as determined by the detectMobile function or has win64 or macintel in its device string then call [updateServerTime](https://github.com/search?q=repo%3Aiiab%2Fiiab-admin-console%20updateServerTime&type=code)
>
> 3. updateServerTime calls a [php script](https://github.com/iiab/iiab-admin-console/blob/master/roles/js-menu/files/menu-files/services/set-server-time.php) that checks to see if the time has been set with NTP and if so returns. It then checks to **see if the UTC version of the system time is more than 2 minutes behind the browser's UTC time** and if so it uses date to set the clock to the browser's UTC time and fake-hwclock to save it.

Reply:
I really don't understand why "elapsed time" calculations became the central point of this discussion. This does only one thing, convert naive datetime objects into timezone-aware datetime-objects, nothing else. The "elapsed time" calculations were always there and we can see their usefullness in the "tasks" page.

Reply:
Deeper Question: what _should_ be the goal of this PR?

If the answer is safer time handling, that might be achieved in ~5 main ways:

1. Disregard the local machine's TZ setting (time zone setting can and commonly does change during IIAB provisioning!)
2. Disregard twice-annual daylight savings time changes (that occur in many time zones!)
3. Disregard machine time changes triggered by NTP
4. Disregard machine time changes triggered by nearby mobile devices whose time (according to their browser headers) is > 2min ahead of IIAB machine's time
5. Disregard machine time changes triggered by sysadmin manual intervention

RECAP: So far this PR transitions to using the local machine's best guess at UTC time — which helps solve 1. and 2.

(But debatably this PR might not be quite enough to address considerations 3., 4. and 5. ?)

Reply:
Quick RECAP of @codewiz's suggestion from ~2 days ago:

Python 3.3's `time.monotonic()` should at least be considered for the "Run Time" column of "Tasks" view — and any other situations where "elapsed time" (or similar) are important... ?

- [time.monotonic()](https://docs.python.org/3/library/time.html#time.monotonic) ([PEP 418](https://peps.python.org/pep-0418/#time-monotonic))
- Linux CLI command `cat /proc/uptime`
- Linux CLI command `uptime`

![image](https://github.com/iiab/calibre-web/assets/2458907/2eccfcb6-eded-429b-a12e-3eed522dde3e)